components/brave_wallet/browser/asset_discovery_manager.h

```diff
@@ -0,0 +1,113 @@
+
+class BraveWalletService;
+class JsonRpcService;
+
+class AssetDiscoveryManager : public mojom::KeyringServiceObserver {
+ public:
+  AssetDiscoveryManager(BraveWalletService* wallet_service,
+                        JsonRpcService* json_rpc_service,
+                        PrefService* prefs);
+
+  AssetDiscoveryManager(const AssetDiscoveryManager&) = delete;
+  AssetDiscoveryManager& operator=(AssetDiscoveryManager&) = delete;
+  ~AssetDiscoveryManager() override;
+
+  // KeyringServiceObserver
+  void KeyringCreated(const std::string& keyring_id) override {}
+  void KeyringRestored(const std::string& keyring_id) override {}
+  void KeyringReset() override {}
+  void Locked() override {}
+  void Unlocked() override {}
+  void BackedUp() override {}
+  void AccountsChanged() override {}
+  void AccountsAdded(mojom::CoinType coin,
+                     const std::vector<std::string>& addresses) override;
+  void AutoLockMinutesChanged() override {}
+  void SelectedAccountChanged(mojom::CoinType coin) override {}
+
+  using APIRequestResult = api_request_helper::APIRequestResult;
+  using EthGetLogsCallback =
+      base::OnceCallback<void(const std::vector<Log>& logs,
+                              mojom::ProviderError error,
+                              const std::string& error_message)>;
+  void DiscoverAssets(const std::string& chain_id,
+                      mojom::CoinType coin,
+                      const std::vector<std::string>& account_addresses,
+                      bool update_prefs,
+                      const std::string& from_block,
+                      const std::string& to_block,
+                      const std::vector<std::string>& supported_chain_ids);
+
+  void OnGetAllTokensDiscoverAssets(
+      const std::string& chain_id,
+      const std::vector<std::string>& account_addresses,
+      std::vector<mojom::BlockchainTokenPtr> user_assets,
+      bool update_prefs,
+      const std::string& from_block,
+      const std::string& to_block,
+      std::vector<mojom::BlockchainTokenPtr> token_list);
+
+  void OnGetTransferLogs(
+      base::flat_map<std::string, mojom::BlockchainTokenPtr>& tokens_to_search,
+      bool update_next_asset_discovery_from_block,
+      const std::string& chain_id,
+      const std::vector<Log>& logs,
+      mojom::ProviderError error,
+      const std::string& error_message);
+
+  void CompleteDiscoverAssets(
+      const std::string& chain_id,
+      std::vector<mojom::BlockchainTokenPtr> discovered_assets,
+      mojom::ProviderError error,
+      const std::string& error_message);
+
+  // Triggered by when KeyringService emits AccountsAdded event.
+  // Rate limits will be ignored, and eth_getLogs query
+  // will run against all blocks, "earliest" to "latest".
+  void DiscoverAssetsOnAllSupportedChains(
+      const std::vector<std::string>& account_addresses,
+      const std::vector<std::string>& supported_chain_ids =
+          GetAssetDiscoverySupportedChains());
```

Comment:
Why do we need to have this supported_chain_ids?
The method is already called all supported chains.
We should make supported_chain list and GetAssetDiscoverySupportedChains() internal to asset discovery manager.

Reply:
In production this should always use the GetAssetDiscoverySupportChains(), but by making this an optional parameter I am able to decouple the supported chain ID list used in the test cases.

I wrote a bunch of test cases that ensure the DiscoverAssetsOnAllSupportedChains runs for multiple chain IDs (and only those chain IDs) as expected, but this will probably be merged with just mainnet supported, so I wanted to find a way to preserve those tests. Now chain IDs we can use in tests do not depend on that list.

https://github.com/brave/brave-core/pull/15707/commits/104518e182aa9da91ec0746929d56653cfd7d437

Reply:
You can expose test only methods to set desired chains, I would prefer that rather than pollute this function signatures just for testing.
And again, please move the list and GetAssetDiscoverySupportedChains() from brave_wallet_constants into this class and make them internal, unless it needs to be public for some reasons.